lxd/cluster/member_state.go

```diff
@@ -0,0 +1,103 @@
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/lxc/lxd/lxd/db"
+	"github.com/lxc/lxd/lxd/state"
+	storagePools "github.com/lxc/lxd/lxd/storage"
+	"github.com/lxc/lxd/shared/api"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+// getLoadAvgs returns the host's load averages from /proc/loadavg.
+func getLoadAvgs() ([]float64, error) {
+	loadAvgs := make([]float64, 3)
+
+	loadAvgsBuf, err := os.ReadFile("/proc/loadavg")
+	if err != nil {
+		return nil, err
+	}
+
+	loadAvgFields := strings.Fields(string(loadAvgsBuf))
+
+	loadAvgs[0], err = strconv.ParseFloat(loadAvgFields[0], 64)
+	if err != nil {
+		return nil, err
+	}
+
+	loadAvgs[1], err = strconv.ParseFloat(loadAvgFields[1], 64)
+	if err != nil {
+		return nil, err
+	}
+
+	loadAvgs[2], err = strconv.ParseFloat(loadAvgFields[2], 64)
+	if err != nil {
+		return nil, err
+	}
+
+	return loadAvgs, nil
+}
+
+// MemberState retrieves state information about the cluster member.
+func MemberState(ctx context.Context, s *state.State, memberName string) (*api.ClusterMemberState, error) {
+	var err error
+	var memberState api.ClusterMemberState
+
+	// Get system info.
+	info := unix.Sysinfo_t{}
+	err = unix.Sysinfo(&info)
+	if err != nil {
+		logger.Warn("Failed getting sysinfo", logger.Ctx{"err": err})
+
+		return nil, err
+	}
+
+	// Account for different representations of Sysinfo_t on different architectures.
+	memberState.SysInfo.Uptime = int64(info.Uptime)
+	memberState.SysInfo.TotalRAM = uint64(info.Bufferram)
+	memberState.SysInfo.SharedRAM = uint64(info.Sharedram)
+	memberState.SysInfo.BufferRAM = uint64(info.Bufferram)
+	memberState.SysInfo.FreeRAM = uint64(info.Freeram)
+	memberState.SysInfo.TotalSwap = uint64(info.Totalswap)
+	memberState.SysInfo.FreeSwap = uint64(info.Freeswap)
+
+	memberState.SysInfo.Processes = info.Procs
+	memberState.SysInfo.LoadAverages, err = getLoadAvgs()
```

Comment:
I used the contents of `/proc/loadavg` here rather than the `Loads` array from `Sysinfo_t` so that its already converted to the more familiar float value that we use in the likes of `top`, and also so it works with LXCFS' load simulation if the cluster members are nested (without needing to use the seccomp sysinfo emulation feature).

However I was intending to use the cgroups feature to populate the other fields in a similar vein, but I wasn't sure if that info is available from a top-level cgroup. For instance there doesn't appear to be a top-level PIDNS I can use to get the process count.

Reply:
Yeah, the top level cgroup is a bit of a mess unfortunately. Not much consistency between controllers or between v1 and v2.